core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -364,10 +364,11 @@ abstract class AbstractFetcherThread(name: String,
 
                         // ReplicaDirAlterThread may have removed topicPartition from the partitionStates after processing the partition data
                         if ((validBytes > 0 || currentFetchState.lag.isEmpty) && partitionStates.contains(topicPartition)) {
+                          val lastFetchedEpoch =
+                            if (validBytes > 0) logAppendInfo.lastLeaderEpoch.asScala else currentFetchState.lastFetchedEpoch
```

Comment:
validBytes would be 0 only in the case when no records returned by the fetch request. In this case, `logAppendInfo.lastLeaderEpoch` will be an empty optional. 

We should update the logic here to say use `currentFetchState.lastFetchedEpoch` if `logAppendInfo.lastLeaderEpoch` is empty. This would decouple the logic of using `lastLeaderEpoch` from the validity of bytes.

Reply:
Make sense. Done